lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -424,164 +425,199 @@ ModulePass *llvm::createDxilCondenseResourcesPass() {
-        if (PHINode *PN = dyn_cast<PHINode>(ChildI))
-          PhiEnd = std::remove(PHIs.begin(), PhiEnd, PN);
-        ChildI->eraseFromParent();
       }
-      BB->eraseFromParent();
     }
   }
 
-  unsigned Attempts = PHIs.size();
-  for (unsigned AttemptIdx = 0; AttemptIdx < Attempts; AttemptIdx++) {
+  if (PHIs.empty())
+    return false;
+
+  bool Changed = false;
+
+  SmallVector<Instruction *, 8> DCEWorklist;
+
+  // Try to simplify those PHI's with DVC and collect them in DCEWorklist
+  for (unsigned Attempt = 0, MaxAttempt = PHIs.size(); Attempt < MaxAttempt; Attempt++) {
     bool LocalChanged = false;
-    for (auto It = PHIs.begin(); It != PHIs.end();) {
-      PHINode *PN = *It;
+    for (unsigned i = 0; i < PHIs.size(); i++) {
+      PHINode *PN = PHIs[i];
       if (Value *V = DVC->GetValue(PN)) {
-
-        PHIs.erase(It);
-        AddCleanupValues(PN); // Mark for deletion later
         PN->replaceAllUsesWith(V);
-        Changed = true;
         LocalChanged = true;
-
-        for (unsigned i = 0, C = PN->getNumIncomingValues(); i < C; i++) {
-          Value *IncomingV = PN->getIncomingValue(i);
-          if (IncomingV != V)
-            AddCleanupValues(IncomingV); // Mark the incoming value for deletion later
-        }
+        DCEWorklist.push_back(PN);
+        PHIs.erase(PHIs.begin() + i);
       }
       else {
-        It++;
+        i++;
       }
     }
 
+    Changed |= LocalChanged;
     if (!LocalChanged)
       break;
   }
 
-  // Simple DCE to remove all dependencies of the resource PHI nodes we removed.
-  // This may be a little too agressive
-  for (;;) {
-    bool LocalChanged = false;
-    // Must use a numeric idx instead of an interator, because
-    // we're modifying the array as we go. Iterator gets invalidated
-    // because they're just pointers.
-    for (unsigned Idx = 0; Idx < CleanupValues.size();) {
-      Value *V = CleanupValues[Idx];
-      if (Instruction *I = dyn_cast<Instruction>(V)) {
-        if (I->user_empty()) {
-          // Add dependencies to process
-          for (Value *Op : I->operands()) {
-            AddCleanupValues(Op);
-          }
-          LocalChanged = true;
-          I->eraseFromParent();
-          CleanupValues.erase(CleanupValues.begin() + Idx);
-        }
-        else {
-          Idx++;
-        }
-      }
-      else {
-        CleanupValues.erase(CleanupValues.begin() + Idx);
-      }
+  // Collect Resource GV loads
+  for (GlobalVariable &GV : M.globals()) {
+    if (!hlsl::dxilutil::IsHLSLObjectType(GV.getType()->getPointerElementType()))
+      continue;
+    for (User *U : GV.users()) {
+      if (Instruction *I = dyn_cast<Instruction>(U))
+        DCEWorklist.push_back(I);
     }
+  }
 
-    Changed |= LocalChanged;
-    if (!LocalChanged)
-      break;
+  // Simple DCE
+  while (DCEWorklist.size()) {
+    Instruction *I = DCEWorklist.back();
+    DCEWorklist.pop_back();
+    if (llvm::isInstructionTriviallyDead(I)) {
```

Comment:
Might need to drill through GEP as well.

Reply:
Still think this will happen, but drill down would be in the earlier loop adding GV users, at which point we could only add if trivially dead to save space/time and redundant copies of the same value.